### PR TITLE
DEP: deprecate misc.pilutil functions

### PR DESCRIPTION
**Pull request by rgommers**
The reasons for these deprecations are:
  1. Many of the functions have bugs and unexpected behavior.
  2. There are better alternatives for pretty much every function.

Summary of issues:

- ``imsave`` rescales images (gh-5305)
- ``imresize`` rescales images (gh-4458) and casts to different dtype (gh-6417)
- Many functions suffer from a shape handling bug for (3, 100, 3) RGB and (4, 100, 4) RGBA images (gh-2347)
- ``toimage`` is the root cause of casting and scaling issues, one more separate bug report at gh-2300
- ``imfilter``, ``imrotate`` and ``imshow`` don't have a bug reported against them, but suffer from casting/rescaling as well.

The functions that seem sort of OK:
- ``imread``. Exposed also as ``ndimage.imread``.  Want to move to ``scipy.io``. Also note the discussion in #6242 (comment) about
  ``imsave``. Want to rename to ``imwrite`` and move to ``scipy.io`` (or write a new one without the rescaling issues and put that in ``io``).
- ``fromimage``.  Changes PIL image to ndarray.
- ``bytescale``.  Does some simple scaling of range of an array.

Closes gh-2347
Closes gh-2300
Closes gh-2442
Closes gh-3154
Closes gh-4458
Closes gh-4893
Closes gh-5305
Closes gh-6242
Closes gh-6417
Closes gh-7259
Closes gh-7458

See discussion about introducing scipy.io.imread/imwrite in http://thread.gmane.org/gmane.comp.python.scientific.devel/20063. I didn't actually do that here, seems duplicate with matplotlib, scikit-image, etc.  Can always be done separately if deemed a good idea.  See also PR gh-5458, which makes an attempt at this.

### CODE REVIEW

File: scipy/misc/pilutil.py [hunk 1]
```diff
@@ -29,7 +29,7 @@
            'imrotate', 'imresize', 'imshow', 'imfilter']
 
 
-# Returns a byte-scaled image
+@numpy.deprecate(message="`bytescale` is deprecated in SciPy 1.0.0.")
```
Comment (larsoner) [inline scipy/misc/pilutil.py]: I think two releases is reasonable. So how about "... and will be removed in 1.2.0", assuming others agree.

File: scipy/misc/pilutil.py [hunk 2]
```diff
@@ -161,12 +163,19 @@ def imread(name, flatten=False, mode=None):
...
 
+@numpy.deprecate(message="`imsave` is deprecated in SciPy 1.0.0. "
+                         "Use ``matplotlib.pyplot.imsave`` instead.")
 def imsave(name, arr, format=None):
     """
     Save an array as an image.
 
     This function is only available if Python Imaging Library (PIL) is installed.
 
+    **Warning**: this function uses `bytescale` under the hood to rescale
```
Comment (larsoner) [inline scipy/misc/pilutil.py]: Can use the `.. warning::` directive

**Review by larsoner — COMMENTED** [review]
Otherwise LGTM, thanks for tackling this.

That is the most beautiful "closes" list I've ever seen :)

**Comment by stefanv** [comment]
I'd suggest either adding imageio as a dependency, and exposing their versions of these functions, or  referring users to that packages.  Our aim with skimage is to eventually do the same.

File: scipy/misc/pilutil.py (hunk 1 above)
Comment (rgommers) [reply scipy/misc/pilutil.py]: Two releases is our default policy, so sounds reasonable to me to start stating that explicitly for all deprecations.

**Comment by rgommers** [comment]
Good point @stefanv. I'll just add a reference now. If we decide we need ``scipy.io.imread/imwrite`` after all, then we should indeed implement that as an ``imageio`` dependency.

**Comment by WarrenWeckesser** [comment]
@stefanv wrote:

> I'd suggest either adding imageio as a dependency, and exposing their versions of these functions, or referring users to that packages. Our aim with skimage is to eventually do the same.

I think the latter option is better.   Wrapping an existing python package that appears to have a solid reputation, works with numpy arrays, and is under active development seems like a waste of effort on our part.  It makes much more sense for us to just point users to imageio.

**Comment by rgommers** [comment]
Comments addressed. Also added a commit that removes ``lena``, which has been generating ``RuntimeError``s for the past 3 releases already.

File: scipy/misc/__init__.py [hunk 3]
```diff
@@ -55,7 +60,6 @@
 __all__ = ['who', 'source', 'info', 'doccer', 'pade',
```
Comment (perimosocordiae) [inline scipy/misc/__init__.py]: If we're not importing `doccer`, we should also remove it from `__all__`.

Comment (rgommers) [reply scipy/misc/__init__.py]: Good catch, I got tricked by my ``pyflakes`` checker. Restored.

**Review by larsoner — APPROVED** [review]

**Comment by rgommers** [comment]
OK in it goes, thanks all.

**Pull request merged**